doc/proposals/scorecard-custom-tests-2.md

```diff
@@ -0,0 +1,293 @@
+
+#### Story 3
+
+As a scorecard developer I would like for custom tests and internal scorecard tests to share the same architecture to provide a cleaner scorecard design and implementation. Scorecard internal tests will migrate to the custom test format and therefore be externalized from the scorecard binary itself.
+
+#### Story 4
+
+I am as an Operator developer, would like to be able to use the Scorecard feature to test the [OLM Bundles](https://github.com/operator-framework/operator-registry/tree/v1.5.3#manifest-format) created for my project.  This story changes Scorcard to require a bundle image name be passed as a configuration CLI flag.  Scorecard will look for example CRs within the bundle's CSV instead of the scorecard configuration.  Scorecard will look for CRDs within the bundle's CSV instead of the scorecard configuration as part of this story.
+
+#### Story 5
+
+I am as an Operator developer, will require the ability to include scorecard confgration file and scorecard test pod manifests into an operator bundle image. 
+
+#### Story 6 (optional)
+
+Scorecard would support the creation of a Kuttl test image.  This custom test would execute kuttl to perform testing enabling scorecard users a means to write tests using kuttl but have the result integrated into scorecard results.  
+
+## Design Details
+
+### Custom Test Definition
+
+* A custom test is any sort of test logic required, written in any programming language that satifies the scorecard test requirements for output format and execution.
+
+### Custom Test Components
+
+This proposal outlines the end-user facing components that make up a custom test including:
+
+#### Test Image
+
+A custom test is based on a container image, that image is created by
+the end-user using whatever tools they want to use.
+
+The test image must produce container log output that conforms to the scorecard Test Result output.  The output is an array of Test Results, so a test image could contain multiple tests if necessary.  Here is a sample of test output in the required format:
+```
+    [
+      {
+	      "name": "Spec fields with descriptors",
+	      "description": "All spec fields have matching descriptors in the CSV",
+	      "state": "fail",
+	      "suggestions": [
+		"Add a spec descriptor for size"
+	      ],
+	      "crname": "example-memcached"
+      }, 
+      {
+	      "name": "Some other test",
+	      "description": "Another test on the CSV",
+	      "state": "pass",
+	      "suggestions": [
+	      ],
+	      "crname": "example-memcached"
+      },
+    ]
+```
+
+The test image will be executed as a Pod by the scorecard with a restart policy of `never`.
+
+When executing a scorecard test, scorecard will create an environment variable in the test container that holds the name of the CR being tested, this allows the test writer a means of knowing which CR is being created by the scorecard as part of this test execution.
```

Comment:
My initial instinct here is that the test image would be responsible for doing whatever setup is necessary from the files it cares about from the bundle.

So if a scorecard test wants to deploy a CR and then make some assertions about what happens as a result, great! But if a scorecard test wants to do a static validation of the bundle, then there's no need to create a CR for that test.

Moreover, I think creating one or more CRs is a step that test images _must_ do because some custom tests may require a complex setup of CRs. For example, this test scenario should be possible: 
1. Create a data CR with some configuration about DB credentials
2. Create an app CR to run an instance of a DB
3. Check that the DB is healthy
4. Populate some data in the DB
5. Create a CR that triggers a backup of the DB
6. Check that the backup contains the populated data.  

None of this is possible with a simplistic single CR test scenario, so I think we leave the test setup up to the test image. All scorecard would do at that point is fetch the bundle, extract the configuration from the bundle about what tests to run, make the bundle contents available to the test containers, run each of the tests, and capture and report their results.

Reply:
sounds reasonable to remove the CR creation from scorecard (the way it works today), and instead go with this approach, that being to let the tests iterate thru the CRs, creating the CRs as they wish.  I suppose the scorecard might still need to clean up the CRs after a test execution still?  or do you think the tests should honor cleaning up after themselves as part of their logic and we don't assign that to the scorecard at all?

Reply:
I think scorecard will likely create a namespace to run the pod for the test image and will delete that namespace after the test completes. So anything in that namespace would get automatically cleaned up, but I think the test image should do its own setup/cleanup if it needs anything more than that.

Reply:
ah ok, I'd not counted on scorcard handling namespace creation/deletion, but if that feature is present it simplifies cleanup of test resources for sure.  Probably a new story for handling the namespaces is necessary,  I'll add it in the next commit.

Reply:
I updated the scorecard processing/workflow section to show it creating and deleting namespaces as it interates thru the tests, removing the CR creation.

Reply:
Do you think it makes sense for scorecard to use the namespace from the environment (e.g. `KUBECONFIG` or `~/.kube/config`, etc.)?

I can see an argument for that too. That would simplify the scorecard execution side of things.

Another thing I just thought of is that test containers will need a service account that gives them enough privileges to execute the test. We need to figure out how that should work.

Reply:
Should not it work as operator-sdk run --local?
Then, should not it have an --operator-namespace and --watch-namespace flags?
Otherwise how would work the operator cluster-scope scenario? @joelanford

Reply:
for service account's in the test pods, we could either a) use the operator's service account which would probably be more privs than necessary, or b) look for a role/binding/SA using a naming convention perhaps (e.g. test-role.yaml, test-role-binding.yaml, etc.) and then create the SA/role/binding on the fly before the tests are run.  Probably option 'b' is suitable since it means the test writer is responsible for creating exactly the SA/role that their tests need, scorecard would just create them as necessary.  For option 'b', either a naming convention to find the security manifests or a CLI flag/config option would tell scorecard where those are to be found.

Not having scorecard deal with namespace creation/deletion greatly simplifies the scorecard implementation.  I'm not sure how scorecard would deal with setting up n-number of namespaces based on varying operator configurations, that could get tricky to figure out and setup correctly.  I'd probably narrow down what scorecard could do here, that would be to work with a single namespace, one currently created, and one that both the operator and the tests could run in at the same time.  Cleanup of resources would be within that namespace instead of depending on namespace deletion for cleanup.

Reply:
What if we went with:
1. We deploy the test images in the same namespace as the operator
2. We use the same service account that the operator is using

It stands to reason that the operator's service account should be pretty close to the superset of rules that a test needs. If that assumption doesn't hold, we can add more flexibility for custom roles later?

Reply:
:+1: